### DISCUSSION TITLE
Better Deprecation Standardization

### DISCUSSION THREAD

**Discussion by Aterxerxes** [npm/feedback#415] [Registry] [answered]
Problem Statement

There is ambiguity in deprecating items and as to the reason why.
Further, there should be the ability to at least (pseudo)enforce updates for private repositories.
Please note that **one** of the reasons why I'm worried about standardization in this area, is that semantic versioning is seen as a bit of a joke in 2+ organizations where I've worked.
In other words, we only use a straight version number and never allow for automatic updating based on patch, or minor versions.

Potential Solution

Add a "deprecated" field to the package.json file that can be filled out like the following:
```
{
  "name": "@fake/faux",
  "version": "3.2.1",
  "author": "Me <me@you.com>",
  "private": true,
  "main": "main.js",
  "description": "Fakening the faux non-deplume.",
  "scripts": {
    "start": "node main.js"
  },
  "dependencies": {
    "vue": "2.6.11"
  },
  "deprecated": [{
      "criticality": "patch",
      "message": "Calculations were incorrect in this version, please update.",
      "from": [
        "4.3.2"
      ],
      "to": [
        "4.3.3"
      ]
    },
    {
      "criticality": "minor",
      "message": "Added some crucial functionality",
      "from": [
        "4.1.6"
      ],
      "to": [
        "4.2.0"
      ]
    },
    {
      "criticality": "major",
      "message": "Attempts were made to patch 3.x to compliance, but there were simply too many breaking changes. We decided to deprecate 3.x. See the link for more info.",
      "link": "https://totallyfakedomain.io/totallyfakepath.html",
      "from": [
        "3.0.0-3.92.3"
      ],
      "to": [
        "4.0.0"
      ]
    },
    {
      "criticality": "critical",
      "message": "Security vulnerability was found.",
      "from": [
        "4.8.3"
      ],
      "to": [
        "4.8.4"
      ]
    },
    {
      "criticality": "mandatory",
      "message": "All employees must move to the new version. C'mon people it's been like 2 years, we shouldn't have any 3.x or 4.x versions out there anymore!",
      "from": [
        "0.0.1-4.8.3"
      ],
      "to": [
        "5.1.2"
      ]
    }
  ]
}
```
Explanation of deprecated fields:
  criticality: Contains some NPM defined criticalities and allows usage of custom fields that could be operated upon.
  message: A message to display if an installed version matches the "from" field.
  from: Which versions are affected by the deprecation. A range or exact match.
  to: Which version they should go to. Could also be a version expression like "5.*" or "5.1.*".

The items I've added hopefully give you an idea of some of the scenarios I'm worried about.
While this doesnt' **directly** solve all problems, it can be used for finer grained control and messaging for some of the issues I've seen throughout my development time.
This would also help some other development teams across the world. Small and large.
I would definitely love to hear some feedback as I'm sure that this is just a good starting point, and not the eventual destination we will arrive at.
Also, I think that some of the naming could use a bit of TLC.

**Comment by ljharb** [comment]
This seems like it should go in https://github.com/npm/rfcs instead of here, but, thoughts below:

---
Making it a package.json field would require publishing a new version in order to change deprecations. Also, which would be the canonical one? The "latest" tag, or the latest semver version, or the latest chronologically published? I could make a strong argument for each of them.

"semantic versioning is seen as a bit of a joke" is a pretty strange comment given that it drives the largest package ecosystem on the planet, so I'm not sure what to do with that one.

`npm deprecate` already allows a package author to precisely deprecate a version (or a range) with as explicit a description as they desire. The only thing that seems different in this suggestion is the "to" - ie, the ability to suggest updating a version. However, wouldn't the "to" always be "the closest un-deprecated version, or higher"?

**Reply by Aterxerxes** [reply]
@ljharb 
Excellent points.

I do want to address each of your comments:
Making it a package.json would require a new release and wouldn't necessarily be the right way to go. Totally valid point.

My comment about "semantic versioning is seen as a bit of a joke" was valid in the circles where I've worked over the course of my career, but I definitely recognize that it is valid in other circles. It just wasn't by us, because the need for stability was high, and some of the dependencies didn't follow semantic versioning as they ought to. (i.e. patches and minor don't break existing functionality).
So yeah, it does drive the largest package ecosystem on the planet, but wasn't ever utilized by the orgs I worked at. So it was a joke there. It's totally valid that some people use it successfully.

`npm deprecate` does already exist. I am, however, trying to add another layer on top of the already existing deprecation functionality that enhances it to the point where we could **hard** enforce moving to a new version if there's a good reason for it. With an override that is **not** the default behavior.
I get that forcing movement to a particular version is not something that sits well, but there are valid use cases for it.

Also, thank you for responding. I appreciate the points you have brought up.
Hopefully we'll get to the heart of the matter well with this discussion.

**Comment by MylesBorins** [comment, marked as answer]
@Aterxerxes thanks for opening the discussion. The next step for this type of proposal would be to open an issue over at https://github.com/npm/rfcs to get direct feedback from the CLI team.
